src/core/thread/energy_scan_server.cpp

```diff
@@ -132,16 +143,22 @@ void EnergyScanServer::HandleScanResult(Mac::EnergyScanResult *aResult, void *aC
 
 void EnergyScanServer::HandleScanResult(Mac::EnergyScanResult *aResult)
 {
-    VerifyOrExit(mActive);
+    VerifyOrExit(mReportMessage != nullptr);
 
     if (aResult)
     {
-        VerifyOrExit(mScanResultsLength < OPENTHREAD_CONFIG_TMF_ENERGY_SCAN_MAX_RESULTS);
-        mScanResults[mScanResultsLength++] = aResult->mMaxRssi;
+        uint16_t tlvLength = mReportMessage->GetLength() - mEnergyListTlvOffset - sizeof(Tlv);
+
+        if ((tlvLength >= NumericLimits<uint8_t>::kMax) ||
```

Comment:
If we exceed the limit, should we try to send the existing stored results?

Reply:
Good point. Yes. We can send the already collected data. Changed this in new pushed commit. 
Also changed the code so we remember the number of appended scan results and then we can calculate the offset of the TLV length field from it.